webapp/graphite/tags/redis.py

```diff
@@ -0,0 +1,167 @@
+    tags                      # Set of all tags
+    tags:<tag>:series         # Set of paths with entry for tag
+    tags:<tag>:values         # Set of values for tag
+    tags:<tag>:values:<value> # Set of paths matching tag/value
+
+  """
+  def __init__(self):
+    from redis import Redis
+
+    self.r = Redis(host=settings.TAGDB_REDIS_HOST,port=settings.TAGDB_REDIS_PORT,db=settings.TAGDB_REDIS_DB)
+
+  def find_series(self, tags):
+    with self.r.pipeline() as pipe:
+      all_match_empty = True
+
+      for tagspec in tags:
+        m = re.match('^([^;!=]+)(!?=~?)([^;]*)$', tagspec)
+        if m is None:
+          raise ValueError("Invalid tagspec %s" % tagspec)
+
+        tag = m.group(1)
+        operator = m.group(2)
+        spec = m.group(3)
+
+        # find list of values that match the tagspec
+        values = None
+
+        if operator == '=':
+          matches_empty = spec == ''
+
+          values = [spec]
+
+        elif operator == '=~':
+          pattern = re.compile(spec)
+          matches_empty = bool(pattern.match(''))
+
+          values = [value for value in self.r.sscan_iter('tags:' + tag + ':values') if pattern.match(value) is not None]
+
+        elif operator == '!=':
+          matches_empty = spec != ''
+
+          values = [value for value in self.r.sscan_iter('tags:' + tag + ':values') if value != spec]
+
+        elif operator == '!=~':
+          pattern = re.compile(spec)
+          matches_empty = not pattern.match('')
+
+          values = [value for value in self.r.sscan_iter('tags:' + tag + ':values') if pattern.match(value) is None]
+
+        else:
+          raise ValueError("Invalid operator %s" % operator)
+
+        # if we're matching the empty value, union with the set of series that don't have the tag
+        if matches_empty:
+          # store series that don't have this tag in a temp key
+          pipe.sdiffstore('temp:' + tag + ':empty', 'series', 'tags:' + tag + ':series')
+
+          # store union of series without this tag and all series that match the spec in a temp key
+          pipe.sunionstore('temp:' + tagspec, 1 + len(values), 'temp:' + tag + ':empty', *['tags:' + tag + ':values:' + value for value in values])
+        # otherwise only return series for matching values
+        else:
+          # if this query matched no values, just short-circuit since the result of the final intersect will be empty
+          if not values:
+            return []
+
+          # store union of all series that match the spec in a temp key
+          pipe.sunionstore('temp:' + tagspec, len(values), *['tags:' + tag + ':values:' + value for value in values])
+
+        all_match_empty = all_match_empty and matches_empty
+
+      if all_match_empty:
+        raise ValueError("At least one tagspec must not match the empty string")
+
+      # get the final result from the intersection of the temp lists we created above
+      pipe.sinter(*['temp:' + tagspec for tagspec in tags])
```

Comment:
That's nice. I was trying to find a race condition when there are different values/operators, but can't find one